server/src/main/kotlin/fi/oph/ludos/assignment/AssignmentController.kt

```diff
@@ -5,28 +5,48 @@ import fi.oph.ludos.Exam
 import org.springframework.data.crossstore.ChangeSetPersister.NotFoundException
 import org.springframework.http.HttpStatus
 import org.springframework.http.ResponseEntity
+import org.springframework.validation.annotation.Validated
 import org.springframework.web.bind.annotation.*
+import javax.validation.Valid
 
 @RestController
+@Validated
 @RequestMapping("${Constants.API_PREFIX}/assignment")
 class AssignmentController(val service: AssignmentService) {
     @PostMapping("")
-    fun createAssignment(@RequestBody assignment: Assignment): ResponseEntity<out Any> =
-        ResponseEntity.status(HttpStatus.OK).body(service.createAssignment(assignment))
+    fun createAssignment(@Valid @RequestBody assignment: Assignment): ResponseEntity<out Any> = when (assignment) {
+        is SukoAssignmentDtoIn -> {
+            ResponseEntity.status(HttpStatus.OK).body(service.createSukoAssignment(assignment))
+        }
+
+        is PuhviAssignmentDtoIn -> {
+            ResponseEntity.status(HttpStatus.OK).body(service.createPuhviAssignment(assignment))
+        }
+
+        is LdAssignmentDtoIn -> {
+            ResponseEntity.status(HttpStatus.OK).body(service.createLdAssignment(assignment))
+        }
+
+        else -> throw UnknownError("Unreachable")
+    }
+
+    @GetMapping("/oppimaaras")
+    fun getOppimaarasInUse(): List<String> = service.getOppimaarasInUse()
 
     @GetMapping("/{exam}")
     fun getAssignments(
         @PathVariable exam: Exam,
-        @RequestParam(required = false) course: String?,
-        @RequestParam(required = false) assignmentTypeKoodiArvo: String?,
-        @RequestParam(required = false) title: String?,
-        @RequestParam(required = false) language: String?,
-        @RequestParam(required = false) orderBy: String?,
-        @RequestParam(required = false) orderDirection: String?
+        @RequestParam(required = false) oppimaara: String?,
+        @RequestParam(required = false) tehtavatyyppisuko: String?,
+        @RequestParam(required = false) aihe: String?,
+        @RequestParam(required = false) tavoitetaitotaso: String?,
+        @RequestParam(required = false) orderDirection: String?,
+        @RequestParam(required = false) lukuvuosi: String?,
+        @RequestParam(required = false) aine: String?,
+        @RequestParam(required = false) tehtavatyyppipuhvi: String?
```

Comment:
Ehdotuksia:
1) Vois splittaa tän endpointin kolmeksi niin, että jokaisella examilla on oma
2) sit vois olla oma AssignmentFilter-luokka jokaisella, esim. SukoAssignmentFilter, jonka toteutuksessa ei tartte huolehtia muiden exameiden filttereistä
3) Silloin vois käyttää samaa nättiä annotaatiovalidaatiosysteemiä kuin mikä tehtävä-createen ja updateen koodailtiin, ei tartte itte heitellä 400-poikkeuksia.
4) Ja mun mielestä ois nätimpää antaa Springin pulauttaa suoraan Filter-luokka get-parametreista luettelematta niitä itse tässä.

Jotenkin näin:
```
@GetMapping("/suko")
fun getSukoAssignments(@Valid SukoAssignmentFilter sukoAssignmentFilter) = service.getAssignments(Exam.SUKO, sukoAssignmentFilter)
```